crates/extensions/c8y_auth_proxy/src/server.rs

```diff
@@ -0,0 +1,307 @@
+use miette::Context;
+use miette::IntoDiagnostic;
+use reqwest::Method;
+use reqwest::StatusCode;
+use std::fmt;
+use std::net::IpAddr;
+use std::sync::Arc;
+use tracing::info;
+
+type AxumServer = hyper::Server<AddrIncoming, IntoMakeService<Router>>;
+
+pub struct Server {
+    fut: BoxFuture<'static, hyper::Result<()>>,
+}
+
+impl Server {
+    pub(crate) fn try_init(state: AppState, address: IpAddr, port: u16) -> miette::Result<Self> {
+        Ok(Server {
+            fut: try_run_server(address, port, state)?.boxed(),
+        })
+    }
+
+    pub fn wait(self) -> BoxFuture<'static, hyper::Result<()>> {
+        self.fut
+    }
+}
+
+fn try_run_server(address: IpAddr, port: u16, state: AppState) -> miette::Result<AxumServer> {
+    info!("Launching on port {port}");
+    let handle = get(respond_to)
+        .post(respond_to)
+        .put(respond_to)
+        .patch(respond_to)
+        .delete(respond_to)
+        .options(respond_to);
+    let app = Router::new()
+        .route("/c8y", handle.clone())
+        .route("/c8y/*path", handle)
+        .with_state(state);
+    Ok(axum::Server::try_bind(&(address, port).into())
+        .into_diagnostic()
+        .wrap_err_with(|| format!("binding to port {port}"))?
+        .serve(app.into_make_service()))
+}
+
+#[derive(Clone)]
+pub(crate) struct AppState {
+    pub target_host: Arc<str>,
+    pub token_manager: SharedTokenManager,
+}
+
+impl FromRef<AppState> for TargetHost {
+    fn from_ref(input: &AppState) -> Self {
+        Self(input.target_host.clone())
+    }
+}
+
+impl FromRef<AppState> for SharedTokenManager {
+    fn from_ref(input: &AppState) -> Self {
+        input.token_manager.clone()
+    }
+}
+
+#[derive(Clone)]
+struct TargetHost(Arc<str>);
+
+impl fmt::Display for TargetHost {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+async fn respond_to(
+    State(TargetHost(host)): State<TargetHost>,
+    retrieve_token: State<SharedTokenManager>,
+    path: Option<Path<String>>,
+    method: Method,
+    headers: HeaderMap<HeaderValue>,
+    small_body: crate::body::PossiblySmallBody,
+) -> Response {
+    let path = match &path {
+        Some(Path(p)) => p.as_str(),
+        None => "",
+    };
+
+    if path.ends_with(".js") || path.starts_with("apps/") {
+        return Response::builder()
+            .status(StatusCode::FORBIDDEN)
+            .body(<_>::default())
+            .unwrap();
+    }
+    let destination = format!("{host}/{path}",);
+
+    let mut token = retrieve_token.not_matching(None).await;
+
+    let client = reqwest::Client::new();
+    let (body, body_clone) = small_body.try_clone();
+    let send_request = |body, token| {
+        client
+            .request(method.to_owned(), destination.clone())
+            .headers(headers.clone())
+            .bearer_auth(&token)
+            .body(body)
+            .send()
+    };
+    let mut res = send_request(body, token.clone()).await.unwrap();
```

Comment:
When can this unwrap go wrong?

There are many unwraps in this method and it's not clear which are safe and which are not.

Reply:
I should adjust this to be more explicit, but this should only fail in the case where the request fails entirely (e.g. connection refused, DNS errors, etc.), which will be turned into a 500 Internal Server Error response when the handler panics.

Reply:
>  which will be turned into a 500 Internal Server Error response when the handler panics.

That's the point I was missing.